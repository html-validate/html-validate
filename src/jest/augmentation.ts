/** module augmentation: jest.d.ts */
import { type ConfigData } from "../config";
import { type Message } from "../message";

declare global {
	/* eslint-disable-next-line @typescript-eslint/no-namespace -- module augmentation */
	namespace jest {
		/* eslint-disable-next-line @typescript-eslint/no-empty-object-type, @typescript-eslint/no-unused-vars -- to match jest declaration */
		interface Matchers<R, T = {}> {
			toBeValid(): R;
			toBeInvalid(): R;
			toHaveError(error: Partial<Message>): R;
			toHaveError(ruleId: string, message: string, context?: any): R;
			toHaveErrors(errors: Array<[string, string] | Record<string, unknown>>): R;

			/**
			 * Validate string or HTMLElement.
			 *
			 * Test passes if result is valid.
			 *
			 * @param config - Optional HTML-Validate configuration object.
			 * @param filename - Optional filename used when matching transformer and
			 * loading configuration.
			 */
			toHTMLValidate(filename?: string): R;
			toHTMLValidate(config: ConfigData, filename?: string): R;
			toHTMLValidate(error: Partial<Message>, filename?: string): R;
			toHTMLValidate(error: Partial<Message>, config: ConfigData, filename?: string): R;

			/**
			 * Writes out the given [[Report]] using codeframe formatter and compares
			 * with snapshot.
			 */
			toMatchCodeframe(snapshot?: string): R;

			/**
			 * Writes out the given [[Report]] using codeframe formatter and compares
			 * with inline snapshot.
			 */
			toMatchInlineCodeframe(snapshot?: string): R;
		}
	}
}
/** module augmentation end */
