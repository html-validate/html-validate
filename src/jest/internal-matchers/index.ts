export { toBeToken } from "./to-be-token";
