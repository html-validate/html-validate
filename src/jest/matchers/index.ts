export { toBeValid } from "./to-be-valid";
export { toBeInvalid } from "./to-be-invalid";
export { toHTMLValidate } from "./to-htmlvalidate";
export { toHaveError } from "./to-have-error";
export { toHaveErrors } from "./to-have-errors";
export { toMatchCodeframe } from "./to-match-codeframe";
export { toMatchInlineCodeframe } from "./to-match-inline-codeframe";
