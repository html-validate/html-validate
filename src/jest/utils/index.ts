export { diff } from "./diff";
export { type MaybeAsyncCallback, diverge } from "./diverge";
export { flattenMessages } from "./flatten-messages";
export { isThenable } from "./is-thenable";
export { type MatcherContext } from "./matcher-context";
export { type MatcherExpect } from "./matcher-expect";
export { type MatcherResult } from "./matcher-result";
export { stripAnsi } from "./strip-ansi";
