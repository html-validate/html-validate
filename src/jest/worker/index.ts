export { type ValidateStringFn } from "./worker";
export { createSyncFn } from "./create-sync-fn";
export { default as workerPath } from "./worker.ts?worker&url";
