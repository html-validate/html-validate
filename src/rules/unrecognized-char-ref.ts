import { Location, sliceLocation } from "../context";
import { HtmlElement, NodeType } from "../dom";
import { AttributeEvent, ElementReadyEvent } from "../event";
import { Rule, RuleDocumentation, SchemaObject, ruleDocumentationUrl } from "../rule";
import entities from "../elements/entities.json";

export interface RuleContext {
	entity: string;
	terminated: boolean;
}

interface RuleOptions {
	ignoreCase: boolean;
	requireSemicolon: boolean;
}

interface EntityMatch {
	/** original regex match object */
	match: RegExpMatchArray;
	/** normalized character reference */
	entity: string;
	/** raw/original character reference (that is, exactly as the user entered it) */
	raw: string;
	/** `true` if the character reference was terminated with a `;` */
	terminated: boolean;
}

const defaults: RuleOptions = {
	ignoreCase: false,
	requireSemicolon: true,
};

const regexp = /&(?:[a-z0-9]+|#x?[0-9a-f]+)(;|[^a-z0-9]|$)/gi;
const lowercaseEntities = entities.map((it) => it.toLowerCase());

function isNumerical(entity: string): boolean {
	return entity.startsWith("&#");
}

function getLocation(
	location: Location | null,
	entity: string,
	match: RegExpMatchArray
): Location | null {
	/* istanbul ignore next: never happens in practive */
	const index = match.index ?? 0;
	return sliceLocation(location, index, index + entity.length);
}

function getDescription(context: RuleContext | undefined, options: RuleOptions): string {
	const url = "https://html.spec.whatwg.org/multipage/named-characters.html";
	let message: string;
	if (context) {
		if (context.terminated) {
			message = `Unrecognized character reference \`${context.entity}\`.`;
		} else {
			message = `Character reference \`${context.entity}\` must be terminated by a semicolon.`;
		}
	} else {
		message = `Unrecognized character reference.`;
	}

	return [
		message,
		`HTML5 defines a set of [valid character references](${url}) but this is not a valid one.`,
		"",
		"Ensure that:",
		"",
		"1. The character is one of the listed names.",
		...(options.ignoreCase ? [] : ["1. The case is correct (names are case sensitive)."]),
		...(options.requireSemicolon ? ["1. The name is terminated with a `;`."] : []),
	].join("\n");
}

export default class UnknownCharReference extends Rule<RuleContext, RuleOptions> {
	public constructor(options: Partial<RuleOptions>) {
		super({ ...defaults, ...options });
	}

	public static schema(): SchemaObject {
		return {
			ignoreCase: {
				type: "boolean",
			},
			requireSemicolon: {
				type: "boolean",
			},
		};
	}

	public documentation(context?: RuleContext): RuleDocumentation {
		return {
			description: getDescription(context, this.options),
			url: ruleDocumentationUrl(__filename),
		};
	}

	public setup(): void {
		this.on("element:ready", (event: ElementReadyEvent) => {
			const node = event.target;

			/* only iterate over direct descendants */
			for (const child of node.childNodes) {
				if (child.nodeType !== NodeType.TEXT_NODE) {
					continue;
				}
				this.findCharacterReferences(node, child.textContent, child.location, {
					isAttribute: false,
				});
			}
		});

		this.on("attr", (event: AttributeEvent) => {
			/* boolean attributes has no value so nothing to validate */
			if (!event.value) {
				return;
			}

			this.findCharacterReferences(event.target, event.value.toString(), event.valueLocation, {
				isAttribute: true,
			});
		});
	}

	private get entities(): string[] {
		if (this.options.ignoreCase) {
			return lowercaseEntities;
		} else {
			return entities;
		}
	}

	private findCharacterReferences(
		node: HtmlElement,
		text: string,
		location: Location | null,
		{ isAttribute }: { isAttribute: boolean }
	): void {
		const isQuerystring = isAttribute && text.includes("?");
		for (const match of this.getMatches(text)) {
			this.validateCharacterReference(node, location, match, { isQuerystring });
		}
	}

	private validateCharacterReference(
		node: HtmlElement,
		location: Location | null,
		foobar: EntityMatch,
		{ isQuerystring }: { isQuerystring: boolean }
	): void {
		const { requireSemicolon } = this.options;
		const { match, entity, raw, terminated } = foobar;

		/* assume numeric entities are valid for now */
		if (isNumerical(entity)) {
			return;
		}

		/* special case: when attributes use query parameters we skip checking
		 * unterminated attributes */
		if (isQuerystring && !terminated) {
			return;
		}

		const found = this.entities.includes(entity);

		/* ignore if this is a known character reference name */
		if (found && (terminated || !requireSemicolon)) {
			return;
		}

		if (found && !terminated) {
			const entityLocation = getLocation(location, entity, match);
			const message = `Character reference "{{ entity }}" must be terminated by a semicolon`;
			const context: RuleContext = {
				entity: raw,
				terminated: false,
			};
			this.report(node, message, entityLocation, context);
			return;
		}

		const entityLocation = getLocation(location, entity, match);
		const message = `Unrecognized character reference "{{ entity }}"`;
		const context: RuleContext = {
			entity: raw,
			terminated: true,
		};
		this.report(node, message, entityLocation, context);
	}

	private *getMatches(text: string): IterableIterator<EntityMatch> {
		let match: RegExpMatchArray | null;
		do {
			match = regexp.exec(text);
			if (match) {
				const terminator = match[1]; // === ";" ? match[1] : "";
				const terminated = terminator === ";";
				const needSlice = terminator !== ";" && terminator.length > 0;
				const entity = needSlice ? match[0].slice(0, -1) : match[0];
				if (this.options.ignoreCase) {
					yield { match, entity: entity.toLowerCase(), raw: entity, terminated };
				} else {
					yield { match, entity, raw: entity, terminated };
				}
			}
		} while (match);
	}
}
