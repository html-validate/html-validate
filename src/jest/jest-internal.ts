import "./jest";

import { Location } from "../context";
import { TokenType } from "../lexer";
import { toBeToken } from "./internal-matchers";

interface TokenMatcher {
	type: TokenType;
	location?: Partial<Location>;
	data?: any;
}

declare global {
	/* eslint-disable-next-line @typescript-eslint/no-namespace */
	namespace jest {
		/* eslint-disable-next-line @typescript-eslint/ban-types, @typescript-eslint/no-unused-vars */
		interface Matchers<R, T = {}> {
			toBeToken(expected: TokenMatcher): R;
		}
	}
}

expect.extend({
	toBeToken,
});
