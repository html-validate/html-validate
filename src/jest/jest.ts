import "./augmentation";

import {
	toBeValid,
	toBeInvalid,
	toHTMLValidate,
	toHaveError,
	toHaveErrors,
	toMatchCodeframe,
	toMatchInlineCodeframe,
} from "./matchers";
import { diff } from "./utils";

expect.extend({
	toBeValid: toBeValid(),
	toBeInvalid: toBeInvalid(),
	toHTMLValidate: toHTMLValidate(expect, diff),
	toHaveError: toHaveError(expect, diff),
	toHaveErrors: toHaveErrors(expect, diff),
	toMatchCodeframe: toMatchCodeframe(),
	toMatchInlineCodeframe: toMatchInlineCodeframe(),
});
