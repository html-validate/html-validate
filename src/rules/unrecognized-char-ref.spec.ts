import HtmlValidate from "../htmlvalidate";
import "../jest";
import { RuleContext } from "./unrecognized-char-ref";

describe("rule unrecognized-char-ref", () => {
	let htmlvalidate: HtmlValidate;

	beforeEach(() => {
		htmlvalidate = new HtmlValidate({
			rules: { "unrecognized-char-ref": "error" },
		});
	});

	describe("text content", () => {
		it("should not report error for valid character reference", () => {
			expect.assertions(1);
			const markup = /* HTML */ ` <p>&amp;</p> `;
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});

		it("should not report error for raw ampersand", () => {
			expect.assertions(1);
			const markup = /* HTML */ ` <p>&</p> `;
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});

		it("should report error when for invalid character reference", () => {
			expect.assertions(2);
			const markup = /* HTML */ `
				<p>&spam;</p>
			`;
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeInvalid();
			expect(report).toMatchInlineCodeframe(`
				"error: Unrecognized character reference "&spam;" (unrecognized-char-ref) at inline:2:8:
				  1 |
				> 2 | 				<p>&spam;</p>
				    | 				   ^^^^^^
				  3 |
				Selector: p"
			`);
		});

		it("should handle multiple entities", () => {
			expect.assertions(2);
			const markup = /* HTML */ `
				<p>&foo; &bar; &baz;</p>
			`;
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeInvalid();
			expect(report).toMatchInlineCodeframe(`
				"error: Unrecognized character reference "&foo;" (unrecognized-char-ref) at inline:2:8:
				  1 |
				> 2 | 				<p>&foo; &bar; &baz;</p>
				    | 				   ^^^^^
				  3 |
				Selector: p
				error: Unrecognized character reference "&bar;" (unrecognized-char-ref) at inline:2:14:
				  1 |
				> 2 | 				<p>&foo; &bar; &baz;</p>
				    | 				         ^^^^^
				  3 |
				Selector: p
				error: Unrecognized character reference "&baz;" (unrecognized-char-ref) at inline:2:20:
				  1 |
				> 2 | 				<p>&foo; &bar; &baz;</p>
				    | 				               ^^^^^
				  3 |
				Selector: p"
			`);
		});

		it("should handle nested elements", () => {
			expect.assertions(1);
			const markup = /* HTML */ ` <div><p>&amp;</p></div> `;
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});
	});

	describe("attribute", () => {
		it("should not report error for valid character reference", () => {
			expect.assertions(1);
			const markup = /* HTML */ ` <p id="&amp;&#123;"></p> `;
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});

		it("should not report error for raw ampersand", () => {
			expect.assertions(1);
			const markup = /* HTML */ `
				<a href="?foo&bar"></p>
				<a href="foo?bar&baz"></p>
			`;
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});

		it("should report error when for invalid character reference", () => {
			expect.assertions(2);
			const markup = /* HTML */ `
				<p title="&spam;"></p>
			`;
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeInvalid();
			expect(report).toMatchInlineCodeframe(`
				"error: Unrecognized character reference "&spam;" (unrecognized-char-ref) at inline:2:15:
				  1 |
				> 2 | 				<p title="&spam;"></p>
				    | 				          ^^^^^^
				  3 |
				Selector: p"
			`);
		});

		it("should handle boolean attributes", () => {
			expect.assertions(1);
			const markup = /* HTML */ ` <p id></p> `;
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});
	});

	describe("ignoreCase", () => {
		const markup = /* HTML */ `
			<p id="lowercase">&nbsp;</p>
			<p id="uppercase">&NBSP;</p>
			<p id="mixedcase">&nBSp;</p>
			<p id="pascalcase">&ApplyFunction;</p>
		`;

		it("should be case sensitive by default", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString(markup);
			expect(report).toMatchInlineCodeframe(`
				"error: Unrecognized character reference "&NBSP;" (unrecognized-char-ref) at inline:3:22:
				  1 |
				  2 | 			<p id="lowercase">&nbsp;</p>
				> 3 | 			<p id="uppercase">&NBSP;</p>
				    | 			                  ^^^^^^
				  4 | 			<p id="mixedcase">&nBSp;</p>
				  5 | 			<p id="pascalcase">&ApplyFunction;</p>
				  6 |
				Selector: #uppercase
				error: Unrecognized character reference "&nBSp;" (unrecognized-char-ref) at inline:4:22:
				  2 | 			<p id="lowercase">&nbsp;</p>
				  3 | 			<p id="uppercase">&NBSP;</p>
				> 4 | 			<p id="mixedcase">&nBSp;</p>
				    | 			                  ^^^^^^
				  5 | 			<p id="pascalcase">&ApplyFunction;</p>
				  6 |
				Selector: #mixedcase"
			`);
		});

		it("should ignore case when option is enabled", () => {
			expect.assertions(1);
			htmlvalidate = new HtmlValidate({
				rules: { "unrecognized-char-ref": ["error", { ignoreCase: true }] },
			});
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});

		it("should retain capitalization in error message", () => {
			expect.assertions(2);
			const markup = /* HTML */ ` <p>&UnKnOwN;</p> `;
			htmlvalidate = new HtmlValidate({
				rules: { "unrecognized-char-ref": ["error", { ignoreCase: true }] },
			});
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeInvalid();
			expect(report).toMatchInlineCodeframe(`
				"error: Unrecognized character reference "&UnKnOwN;" (unrecognized-char-ref) at inline:1:5:
				> 1 |  <p>&UnKnOwN;</p>
				    |     ^^^^^^^^^
				Selector: p"
			`);
		});
	});

	describe("requireSemicolon", () => {
		const markup = /* HTML */ `
			<p id="with-semicolon">&copy; &COPY;</p>
			<p id="without-semicolon">&copy &COPY</p>
		`;

		it("should require semicolon by default", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString(markup);
			expect(report).toMatchInlineCodeframe(`
				"error: Character reference "&copy" must be terminated by a semicolon (unrecognized-char-ref) at inline:3:30:
				  1 |
				  2 | 			<p id="with-semicolon">&copy; &COPY;</p>
				> 3 | 			<p id="without-semicolon">&copy &COPY</p>
				    | 			                          ^^^^^
				  4 |
				Selector: #without-semicolon
				error: Character reference "&COPY" must be terminated by a semicolon (unrecognized-char-ref) at inline:3:36:
				  1 |
				  2 | 			<p id="with-semicolon">&copy; &COPY;</p>
				> 3 | 			<p id="without-semicolon">&copy &COPY</p>
				    | 			                                ^^^^^
				  4 |
				Selector: #without-semicolon"
			`);
		});

		it("should allow missing semicolon for entities with legacy compatibility when option is enabled", () => {
			expect.assertions(1);
			htmlvalidate = new HtmlValidate({
				rules: { "unrecognized-char-ref": ["error", { requireSemicolon: false }] },
			});
			const report = htmlvalidate.validateString(markup);
			expect(report).toBeValid();
		});

		it("should always require semicolon for entities without legacy compatibility", () => {
			expect.assertions(1);
			htmlvalidate = new HtmlValidate({
				rules: { "unrecognized-char-ref": ["error", { requireSemicolon: false }] },
			});
			const markup = /* HTML */ ` <p>&star</p> `;
			const report = htmlvalidate.validateString(markup);
			expect(report).toMatchInlineCodeframe(`
				"error: Unrecognized character reference "&star" (unrecognized-char-ref) at inline:1:5:
				> 1 |  <p>&star</p>
				    |     ^^^^^
				Selector: p"
			`);
		});
	});

	it("should contain documentation", () => {
		expect.assertions(1);
		expect(htmlvalidate.getRuleDocumentation("unrecognized-char-ref")).toMatchSnapshot();
	});

	it("should contain contextual documentation (terminated)", () => {
		expect.assertions(1);
		const context: RuleContext = {
			entity: "&spam;",
			terminated: true,
		};
		expect(
			htmlvalidate.getRuleDocumentation("unrecognized-char-ref", null, context)
		).toMatchSnapshot();
	});

	it("should contain contextual documentation (not terminated)", () => {
		expect.assertions(1);
		const context: RuleContext = {
			entity: "&spam;",
			terminated: false,
		};
		expect(
			htmlvalidate.getRuleDocumentation("unrecognized-char-ref", null, context)
		).toMatchSnapshot();
	});
});
