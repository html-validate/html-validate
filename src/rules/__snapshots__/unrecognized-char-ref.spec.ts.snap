// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rule unrecognized-char-ref should contain contextual documentation (not terminated) 1`] = `
{
  "description": "Character reference \`&spam;\` must be terminated by a semicolon.
HTML5 defines a set of [valid character references](https://html.spec.whatwg.org/multipage/named-characters.html) but this is not a valid one.

Ensure that:

1. The character is one of the listed names.
1. The case is correct (names are case sensitive).
1. The name is terminated with a \`;\`.",
  "url": "https://html-validate.org/rules/unrecognized-char-ref.html",
}
`;

exports[`rule unrecognized-char-ref should contain contextual documentation (terminated) 1`] = `
{
  "description": "Unrecognized character reference \`&spam;\`.
HTML5 defines a set of [valid character references](https://html.spec.whatwg.org/multipage/named-characters.html) but this is not a valid one.

Ensure that:

1. The character is one of the listed names.
1. The case is correct (names are case sensitive).
1. The name is terminated with a \`;\`.",
  "url": "https://html-validate.org/rules/unrecognized-char-ref.html",
}
`;

exports[`rule unrecognized-char-ref should contain documentation 1`] = `
{
  "description": "Unrecognized character reference.
HTML5 defines a set of [valid character references](https://html.spec.whatwg.org/multipage/named-characters.html) but this is not a valid one.

Ensure that:

1. The character is one of the listed names.
1. The case is correct (names are case sensitive).
1. The name is terminated with a \`;\`.",
  "url": "https://html-validate.org/rules/unrecognized-char-ref.html",
}
`;
